bin/install

```diff
@@ -14,13 +14,25 @@ install_maven() {
 	build_copy_cleanup $version "$destdir"
 }
 
+# Function to compare versions
+version_lt() {
+	# returns 0 (true) if first version is less than second version, otherwise 1 (false)
+	[ "$(printf '%s\n' "$1" "$2" | sort -V | head -n1)" != "$2" ]
+}
+
 # Download Maven source from Apache
 get_maven() {
 	local version=$1
 	local destdir=$2
 
 	local major=$(echo $version | cut -d '.' -f 1)
 	local base="http://dlcdn.apache.org/maven"
+
+	# Set base URL conditionally based on version
+	if version_lt "$version" "3.8.8"; then
```

Comment:
**suggestion:** if we hardcode this version, we'll need to create a new MR to bump this value every time a version is archived. 

What do you think about changing the download logic to the following:

1. attempt to download the selected version from the main URL http://dlcdn.apache.org/maven 
1. if the selected version can't be downloaded from the main URL, then fallback to the archive URL https://archive.apache.org/dist/maven.
1. if the selected version can't be downloaded from either the main or archive URL, return an error message.

For example:

```shell
download_file() {
	local url=$1
	local dest=$2

	curl -fLC - --retry 3 --retry-delay 3 -o "$dest" "$url"
	return $?
}

get_maven() {
	local version=$1
	local destdir=$2

	local major=$(echo $version | cut -d '.' -f 1)
	local main_url="http://dlcdn.apache.org/maven/maven-$major/$version/binaries/apache-maven-$version-bin.tar.gz"
	local archive_url="https://archive.apache.org/dist/maven/maven-$major/$version/binaries/apache-maven-$version-bin.tar.gz"

	if [[ "$version" == *SNAPSHOT ]]; then
		local snapshot_url="https://repository.apache.org/service/local/artifact/maven/redirect?r=snapshots&g=org.apache.maven&a=apache-maven&v=$version&e=tar.gz&c=bin"
		download_file "$snapshot_url" "$destdir/$version.tar.gz"
		if [ $? -ne 0 ]; then
			ASDF_MAVEN_ERROR="Could not download SNAPSHOT version $version from $snapshot_url. Perhaps the version of Maven you're trying to install does not exist"
			return 1
		fi
		return 0
	fi

	# Attempt to download from the main URL
	download_file "$main_url" "$destdir/$version.tar.gz"
	if [ $? -ne 0 ]; then
		# If the download fails, attempt to download from the archive URL
		download_file "$archive_url" "$destdir/$version.tar.gz"
		if [ $? -ne 0 ]; then
			ASDF_MAVEN_ERROR="Could not download Maven $version from either the main or archive URL. Perhaps the version of Maven you're trying to install does not exist"
			return 1
		fi
	fi
}
```

**note:** I'm not a contributor to this project, so take my suggestion with a grain of salt.

Reply:
I like this approach better - thanks @adamcohen2

Reply:
Updated with your suggestion!